### DISCUSSION TITLE
Inverse kinematics for a 3D human skeleton

### DISCUSSION THREAD

**Discussion by ShadowDawg** [stack-of-tasks/pinocchio#2198] [Q&A]
Hi Pinocchio Community!
I'm currently working on a project that involves calculating the joint angles and velocities for a 3D humanoid skeleton from given joint coordinates and their connections. The dataset I'm working with contains the (x, y, z) coordinates of each joint for specific timesteps, alongside information about how these joints are connected to form the humanoid structure. My goal is to use this data to perform inverse kinematics and calculate joint angles and velocities.
Could anyone suggest where to begin with Pinocchio to accomplish this task?

**Comment by jorisv** [comment]
Hello,

Pinocchio doesn't have an IK module right now. But you can give look at [Pink](https://github.com/stephane-caron/pink/) to achieve that.

**Reply by stephane-caron** [reply]
This discussion is about IK for a human skeleton model, let's not sidetrack it into discussing other types of robot models (which, in this instance, we already discuss in the thread you linked). If you have further questions about Pink, you can ask them in the [Pink issue tracker](https://github.com/stephane-caron/pink/issues).

Also, note that pinging users who were not part of the original discussion (unless you have their consent to do so) can be seen as bad [netiquette](https://en.wiktionary.org/wiki/netiquette). Make sure you ping reasonably to increase your chances of getting useful feedback from fellow users.

**Comment by stephane-caron** [comment]
You can also check out [PlaCo](https://placo.readthedocs.io/en/latest/) and [TSID](https://github.com/stack-of-tasks/tsid), two whole-body inverse kinematics/dynamics libraries based on Pinocchio.
